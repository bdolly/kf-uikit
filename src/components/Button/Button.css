.Button {
  @apply inline-block
        bg-white
        px-4
        py-2
        border
        border-greyBorder
        rounded-lg
        font-bold
        font-title
        text-xs
        text-tealA11y
        uppercase
        leading-tight;

  &:disabled {
    @apply bg-bluegrey
            border
            border-greyBorder
            text-bluegrey-darkest;
    cursor: not-allowed;

    &:hover {
      @apply bg-bluegrey
                border
                border-greyBorder
                text-bluegrey-darkest;
    }
  }

  &:hover {
    @apply bg-lightGrey
            text-tealHover;
  }
}

.Button--large {
  @apply px-20
        py-8
        text-base;
}

.Button--primary {
  @apply bg-tealA11y
        border-tealA11y
        text-white;

  &:hover {
    @apply bg-tealHover
            border-tealHover
            text-white;
  }
}

.Button--secondary {
  @apply bg-purple
        border-purple
        text-white;

  &:hover {
    @apply bg-purpleHover
            border-purpleHover
            text-white;
  }
}
