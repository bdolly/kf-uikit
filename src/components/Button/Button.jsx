import React from 'react';
import propTypes from 'prop-types';
import classes from 'classnames';

/**
 * A Simple rounded button
 */
const Button = ({ size, color, outline, disabled, className, children }) => {

  const buttonClass = classes(
    'Button',
    {[`Button--${size}`]: ['small', 'large'].includes(size)},
    `Button${outline ? '--outline' : ''}--${color}`,
    className
  );
  return (
    <button type='button' className={buttonClass} disabled={disabled}>
      {children}
    </button>
  );
}

Button.propTypes = {
  /** The size of the button */
  size: propTypes.string,
  /** The color of the button. */
  color: propTypes.string,
  /** Whether the button is an outline only, or solid. */
  outline: propTypes.bool,
  /** Whether the button is disabled */
  disabled: propTypes.bool,
  /** Any additional classes to be applied to the button */
  className: propTypes.string,
  /** Children elements. */
  children: propTypes.string,
};

Button.defaultProps = {
  size: 'default',
  color: 'primary',
  outline: false,
  disabled: false,
  className: null,
  children: null,
};

export default Button;
