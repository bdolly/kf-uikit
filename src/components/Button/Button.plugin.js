/* eslint-disable import/no-extraneous-dependencies */
const _ = require('lodash');
const defaultConfig = require('tailwindcss/defaultConfig')();

function defaultOptions() {
  return {
    borderRadius: '1.5rem',
    fontWeight: '600',
    lineHeight: '1.25',
    fontSize: '1rem',
    padding: '.5rem 1rem',
    colors: {
      white: {
        background: defaultConfig.colors.white,
        text: defaultConfig.colors.black,
      },
      black: {
        background: defaultConfig.colors.black,
        text: defaultConfig.colors.white,
      },
    },
    sizes: {
      small: {
        fontSize: '.7rem',
        padding: '.3rem .6rem',
        borderRadius: '1.0rem',
      },
      large: {
        fontSize: '1.25rem',
        padding: '.75rem 1.5rem',
        borderRadius: '1.7rem',
      },
    },
  };
}

module.exports = opts => {
  const options = _.isFunction(opts) ? opts(defaultOptions()) : _.defaults(opts, defaultOptions());

  return ({ addComponents, e }) => {
    addComponents([
      ..._.map(_.omit(options.sizes, 'default'), (sizeOptions, name) => ({
        [`.Button--${e(name)}`]: {
          padding: _.get(sizeOptions, 'padding', options.padding),
          fontSize: _.get(sizeOptions, 'fontSize', options.fontSize),
          fontWeight: _.get(sizeOptions, 'fontWeight', options.fontWeight),
          lineHeight: _.get(sizeOptions, 'lineHeight', options.lineHeight),
          borderRadius: _.get(sizeOptions, 'borderRadius', options.borderRadius),
        },
      })),
    ]);
  };
};
