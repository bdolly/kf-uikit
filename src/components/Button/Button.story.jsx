/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select } from '@storybook/addon-knobs';
import { generateVariantsFor } from '../../utils/propsVariants';
import Button from './Button';
import iconNames from '../Icon/Icons';

const stories = storiesOf('Buttons', module);

stories.add(
  'Button',
  () => {
    const variants = generateVariantsFor({
      size: ['default', 'large'],
      color: ['default', 'primary', 'secondary'],
      disabled: [true, false],
      icon: [null, Object.keys(iconNames)[0]],
    });
    return (
      <section>
        <Button
          className="hide-in-tests"
          size={select('Size', ['default', 'large'], 'default')}
          color={select('Color', ['default', 'primary', 'secondary'], 'default')}
          disabled={boolean('Disabled', false)}
          type="button"
          icon={select('icon', [null, ...Object.keys(iconNames)], null)}
        >
          Click Me
        </Button>
        {process.env.STORYBOOK_PERCY_ENV ? (
          <div>
            {variants.map(props => (
              <Button className="m-20" {...props}>
                Button&nbsp;
                {Object.values(props).join(' ')}
              </Button>
            ))}
          </div>
        ) : null}
      </section>
    );
  },
  {
    info: {
      text: `
        The standard action button for most use cases in the portal.
        `,
    },
  },
);
