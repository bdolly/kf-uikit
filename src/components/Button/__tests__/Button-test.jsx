import React from 'react';
import renderer from 'react-test-renderer';
import Button from '../Button';

['default', 'small', 'large'].map((size) => {
  [true, false].map((disabled) => {
    [true, false].map((outline) => {
      ['primary', 'secondary', 'tertiary'].map((c) => {
        it(`${disabled ? 'disabled ' : ''}${size} ${c} ${outline ? 'outlined ' : ''}button renders correctly`, () => {
            const tree = renderer.create(
                    <Button
                      color={c}
                      size={size}
                      outline={outline}
                      disabled={disabled}
                    >
                      Hello
                    </Button>
                    ).toJSON();
              expect(tree).toMatchSnapshot();
        });
      });
    });
  });
});
