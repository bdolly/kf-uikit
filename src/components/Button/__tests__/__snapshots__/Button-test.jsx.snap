// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="bg-primary hover:bg-primary-dark Button"
  disabled={false}
  type="button"
>
  Hello
</button>
`;
